modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadata.java

```diff
@@ -170,6 +178,54 @@ public boolean sameSnapshot(SnapshotMetadata compare) {
             Objects.equals(baselineNodes(), compare.baselineNodes());
     }
 
+    /**
+     * @param grpId Cache id or cache group id.
+     * @return {@code True} if cache group is encrypted. {@code False} otherwise.
+     */
+    public boolean isCacheGroupEncrypted(int grpId) {
+        Set<Integer> encrGrpIds = this.encrGrpIds;
```

Comment:
to avoid NPE if class member has been nulled after `encrGrpIds != null`

Reply:
But the reference `encrGrpIds` is not synced: reference isn't volatile, method doesn't acquire a lock. So this reference isn't observable for other threads within this method, so this var doesn't make sense. 

Make var volatile, or use lock there.

Reply:
Exactly. Local variable which is not affected by other thred. It guaranties the reference won't change during several reads. Please check https://github.com/apache/ignite/pull/9377 , `EncryptedSnapshotTest.testNPE()`. Try uncomment `//        Set<Integer> encrGrpIds = this.encrGrpIds;` in `SnapshotMetadata.isCacheGroupEncrypted()`

Reply:
This is common technique to avoid NPE to me, introducing local varaible. But in our case, as mentioned, once encrGrpIds is assigned, it won't set to null. We can check it once. But for general purpose I wrote it correct

Reply:
You've initialized `encrGrpIds` with the DCL pattern (that requires the object to be declared with `volatile` keyword, and you missed it). It means that `encrGrpIds` is assigned from concurrent context, but your local var just copy thread local var (_that is not observable from other threads_ and may can't see that `encrGrpIds` is initialized) to local var.

Reply:
Exactly. This brings thread-safety. The test shows it. Makes sense?

Reply:
No encrypted groups ids reside in SnapshotMeta any more. Inactual.